File: Algo/Medium/0153_Find_Minimum_in_Rotated_Sorted_Array/Solution.py
from typing import *
from math import floor

## 2022-11-24
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left = 0
        right = len(nums)-1
        smallest = min(nums[left], nums[right])
        
        while left <= right:
            mid = floor(left + (right - left)/2)
            
            smallest = min(smallest, nums[mid])
            
            if nums[mid] < nums[right]:
                right = mid-1
            else:
                left = mid+1
                
        return smallest

File: Algo/Medium/0153_Find_Minimum_in_Rotated_Sorted_Array/Solution.java
// 2021-08-30
class Solution {
    public int findMin(int[] nums) {
        if (nums.length == 1) return nums[0];
        int left = 0;
        int right = nums.length - 1;
        int lowest = Math.min(nums[left], nums[right]);
        
        while (right - left > 1) {
            int mid = (left + right) / 2;
            lowest = Math.min(nums[mid], lowest);
            if (nums[left] > nums[right]) {
                if (nums[mid] > nums[left]) {
                    left = mid;
                } else {
                    right = mid;
                }
            } else {
                break;
            }
        }
        
        return lowest;
    }
}